test/core/typing.test.ts

```diff
@@ -15,389 +18,661 @@ interface TestContext {
-    const { room } = context;
+      // CHA-T5c
+      it<TestContext>('throws an error if typing.stop is called when the channel is not attached', async (context) => {
+        const { room, realtime } = context;
+        const channel = room.typing.channel;
+        const realtimeChannel = realtime.channels.get(channel.name);
+        vi.spyOn(realtimeChannel, 'publish').mockImplementation(async (): Promise<void> => {});
+        vi.spyOn(room.typing.channel, 'state', 'get').mockReturnValue('attached');
+        await room.typing.keystroke();
+        vi.spyOn(room.typing.channel, 'state', 'get').mockReturnValue('detached');
 
-    // Add a listener
-    const receivedEvents: TypingEvent[] = [];
-    const { unsubscribe } = room.typing.subscribe((event: TypingEvent) => {
-      receivedEvents.push(event);
-    });
+        await expect(room.typing.stop()).rejects.toBeErrorInfoWithCode(50000);
 
-    // Add another
-    const receivedEvents2: TypingEvent[] = [];
-    const { unsubscribe: unsubscribe2 } = room.typing.subscribe((event: TypingEvent) => {
-      receivedEvents2.push(event);
-    });
+        // Check that no messages were sent
+        expect(realtimeChannel.publish).toHaveBeenCalledTimes(1);
+      });
 
-    const channel = context.room.typing.channel;
-    let arrayToReturn = presenceGetResponse(['otherClient']);
-    vi.spyOn(channel.presence, 'get').mockImplementation(() => {
-      return Promise.resolve<Ably.PresenceMessage[]>(arrayToReturn);
-    });
+      it<TestContext>('when stop is called, immediately stops typing', async (context) => {
+        const { realtime, room } = context;
+        const channel = room.typing.channel;
+        const realtimeChannel = realtime.channels.get(channel.name);
 
-    // Emulate a typing event
-    context.emulateBackendPublish({
-      clientId: 'otherClient',
-      action: 'enter',
-    });
+        // If stop is called, it should publish a leave message
+        vi.spyOn(realtimeChannel, 'publish').mockImplementation(async (): Promise<void> => {});
+        vi.spyOn(room.typing.channel, 'state', 'get').mockReturnValue('attached');
 
-    await waitForArrayLength(receivedEvents, 1);
-    expect(channel.presence.get).toBeCalledTimes(1);
+        // Start typing and then immediately stop typing
+        await room.typing.keystroke();
+        await room.typing.stop();
 
-    // Ensure that the listener received the event
-    expect(receivedEvents).toHaveLength(1);
-    expect(receivedEvents[0]).toEqual({
-      currentlyTyping: new Set(['otherClient']),
-    });
+        expect(realtimeChannel.publish).toHaveBeenCalledTimes(2);
+        // Ensure that publish was called with typing.started only once
+        expect(realtimeChannel.publish).toHaveBeenCalledWith(startMessage);
+        // Ensure that publish was called with typing.stopped only once
+        expect(realtimeChannel.publish).toHaveBeenCalledWith(stopMessage);
```

Comment:
Should these be be `toHaveBeenNthCalledWith()` and `toHaveBeenLastCalledWith()` to also ensure start was published before stop?